Library/Homebrew/utils/shell.rb

```diff
@@ -0,0 +1,64 @@
+require 'shellwords'
+
+module Utils
+  module Shell
+    # take a path and heuristically convert it
+    # to a shell, return nil if there's no match
+    def self.path_to_shell(path)
+      # remove leading - from login shell
+      shell_name = File.basename(path).strip_prefix("-")
+      case shell_name
+      when "sh", "bash", "zsh", "csh", "tcsh", "fish", "ksh"
+        shell_name.to_sym()
+      else nil
+      end
+    end
+
+    def self.preferred_shell
+      path_to_shell(ENV.fetch("SHELL", ""))
+    end
+
+    def self.parent_shell
+      path_to_shell(`ps -p #{Process.ppid} -o command=`.chomp)
+    end
+
+    def self.csh_quote(str)
+      # shellquote breaks on csh
+      # put everything in single quotes except
+      # single quotes themselves, which are double quoted
+      # we need to prepend a \ to literal newlines and
+      # ! (history substitution character)
+      substrings = str.split("'", -1)
+      substrings.map { |s| "'" + (s.gsub(/([\n!])/) { '\\' + $1 }) + "'" }.join("\"'\"")
+    end
+
+    # quote values. quoting keys is overkill
+    def self.export_value(shell, key, value)
+      case shell
+      when :bash, :sh, :zsh, :ksh
+        "export #{key}=#{Shellwords.shellescape(value)}"
+      when :fish
+        # fish quoting is mostly Bourne compatible except that
+        # a single quote can be included in a single-quoted string via \'
+        # and a literal \ can be included via \\
+        "set -gx #{key} #{Shellwords.shellescape(value)}"
+      when :tcsh, :csh
+        "setenv #{key} #{csh_quote(value)}"
+      end
+    end
+
+    # return the shell profile file based on users' preferred shell
+    def self.shell_profile
+      case preferred_shell
+      when :bash then "~/.bash_profile"
+      when :zsh then "~/.zshrc"
+      when :ksh then "~/.kshrc"
+      when :tcsh then "~/.tcshrc"
+      when :csh then "~/.cshrc"
+      when :sh then "~/.bash_profile" # /bin/sh is POSIX-mode bash on OS X
+      when :fish then "~/.config/fish/config.fish"
```

Comment:
It would be nice to have these cases sorted by name. Even better (I think) would be to move the mapping into a constant `Hash` and fetch the entries from it. This could look something like this:

```ruby
SHELL_PROFILE_MAP = {
  :bash => "~/.bash_profile",
  :csh => "~/.cshrc",
  :fish => "~/.config/fish/config.fish",
  :ksh => "~/.kshrc",
  :sh => "~/.bash_profile", # /bin/sh is POSIX-mode bash on OS X
  :tcsh => "~/.tcshrc",
  :zsh => "~/.zshrc",
}.freeze

def self.shell_profile
  SHELL_PROFILE_MAP.fetch(preferred_shell, "~/.bash_profile")
end
```